tools/codegen/api/types.py

```diff
@@ -479,5 +480,17 @@ def returns_type(self) -> CType:
     def dispatcher_exprs(self) -> List[Expr]:
         return translate.translate(self.arguments(), dispatcher.arguments(self.func), method=False)
 
+
+# Helper functions
+
+def kernel_signature(f: NativeFunction, backend_index: BackendIndex) -> Union['NativeSignature', 'DispatcherSignature']:
+    # Note [External Backends Follow Dispatcher API]
+    # Kernel signatures for in-tree backends follow the "native" API,
+    # while kernels for out-of-tree backends follow the dispatcher API.
```

Comment:
So, my take is that we **want** to enforce/encourage external backends to all follow the dispatcher API convention if we can: both because (a) any differences between the dispatcher API and the API of the backend kernel results in runtime overhead to convert between the two, and (b) enforcing some uniformity across backends will be easier to manage.

We have a separate native API in-tree, but my understanding is that this just emerged from the fact that our native kernel signatures drifted from the dispatcher API at some point, and introducing a way to translate between the two API's is easier than going through every kernel in the codebase and fixing up its signature.

Compare that to external backends: XLA kernels already follow the dispatcher convention, and it'll be easy to enforce that new backends write kernels that follow the same convention. It seems likely that other backends would follow the same pattern, since we advertise `RegistrationDeclarations.yaml` as the way to tell which operators a backend has to implement, and the names of the ops in that file all follow dispatcher convention.

But you're right, there's still some chance though that there's a backend out there that already wrote a bunch of kernels and didn't follow the dispatcher naming/schema convention. My take is to allow per-backend kernel API's only if we see that somewhere. Another extension I could imagine we might have to do is if the backend gave all of their kernels custom names that didn't follow any of the API's we currently use- we could probably add an option in the yaml to provide a custom kernel name or something, but for the same reason I don't think we'd want that kind of extension unless there's a need for it.

Let me know if that whole argument made sense 😄.

Reply:
So, my take is that we want to enforce/encourage external backends to all follow the dispatcher API convention if we can: both because (a) any differences between the dispatcher API and the API of the backend kernel results in runtime overhead to convert between the two, and (b) enforcing some uniformity across backends will be easier to manage.

We have a separate native API in-tree, but my understanding is that this just emerged from the fact that our native kernel signatures drifted from the dispatcher API at some point, and introducing a way to translate between the two API's is easier than going through every kernel in the codebase and fixing up its signature.

Compare that to external backends: XLA kernels already follow the dispatcher convention, and it'll be easy to enforce that new backends write kernels that follow the same convention. It seems likely that other backends would follow the same pattern, since we advertise RegistrationDeclarations.yaml as the way to tell which operators a backend has to implement, and the names of the ops in that file all follow dispatcher convention.

But you're right, there's still some chance though that there's a backend out there that already wrote a bunch of kernels and didn't follow the dispatcher naming/schema convention. My take is to allow per-backend kernel API's only if we see that somewhere. Another extension I could imagine we might have to do is if the backend gave all of their kernels custom names that didn't follow any of the API's we currently use- we could probably add an option in the yaml to provide a custom kernel name or something, but for the same reason I don't think we'd want that kind of extension unless there's a need for it.

Let me know if you agree with that argument 😄

Reply:
Reading over your comment, @bdhirsh, I'm actually imagining the opposite situation, where it is kernels in PyTorch that are easier to migrate (because we have access to them and can easily change them), and external backend functions that might want to move more slowly because we can't just update them whenever we make changes.  So actually you could imagine an endless stream of versions of signatures, and backends can migrate to the newer versions by opting in. Or I guess we can force them to rip off the bandaid every version release they migrate onto...

Reply:
Yeah that makes perfect sense, and I think you're right that we should proceed as though the native API is legacy - not support it in out-of-tree codebase unless some case comes up that forces the issue. (It might be worth adding a sentence to the comment here giving the high-level color on the native API - otherwise it's pretty easy to read the comment as implying something like "fundamental differences between in tree and out of tree backends have led us to design two different APIs" or whatever.

Reply:
Right. I'm thinking about the following two things as kind of independent (but maybe that's wrong):

**How to ease BC issues for external backends**
This applies to stuff like "we added a new default-able param to `smooth_l1_loss`", or "we renamed an operator from `_{op}` to `{op}`". Right now, any naming/schema changes to any native op will break external backends (their master will fail to build) until they make the corresponding change to their kernel. The codegen gives us a hook into making that experience easier, although I'm not too sure what the design for that would look like. I remember talking with Ed a few weeks ago about some notion of explicit versioning, probably per operator, that a backend would need to specify. One way is for backends to specify "XLA backend has a kernel for version 2 of `at::add`", and to make our codegen aware of how to translate between schema versions. This would put a little more onus on people that modify op schema's in-tree, and doesn't fix all of the BC issues, but it would provide at least some form of shielding backends from BC changes.

The big distinction in my head is that these changes apply on a per-operator basis.

**What naming/schema conventions a backend should abide by**
In my original comment above, I was thinking more around "how does a backend choose what names to give their kernels" or "do they include packed vs. unpacked TensorOptions", which are conventions that I think should apply to all kernels. So, sort of "what is the default naming/schema convention of external backend kernels", which right now in the codegen is the dispatcher API. We could technically allow external backends to abide by some other convention (like our native API). But I was thinking of that as a different problem compared to dealing with BC-breaking operator changes, which needs to be solved separately.